Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,226 @@
+    if (!copiedImage)
+        return nullptr;
+
+    // Compute pattern space transformation.
+    AffineTransform patternSpaceTransform;
+    patternSpaceTransform.translate(tileBoundaries.location());
+    patternSpaceTransform.scale(tileBoundaries.size() / tileImageSize);
+
+    AffineTransform patternTransform = m_attributes->patternTransform();
+    if (!patternTransform.isIdentity())
+        patternSpaceTransform = patternTransform * patternSpaceTransform;
+
+    // Build pattern.
+    return Pattern::create({ copiedImage.releaseNonNull() }, { true, true, patternSpaceTransform });
+}
+
+bool RenderSVGResourcePattern::buildPatternIfNeeded(GraphicsContext& context, const RenderLayerModelObject& targetRenderer)
+{
+    collectPatternAttributesIfNeeded();
+
+    m_pattern = buildPattern(context, targetRenderer);
+
+    if (!m_pattern)
+        return false;
+
+    return true;
+}
+
+bool RenderSVGResourcePattern::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
+        return false;
+
+    if (!buildPatternIfNeeded(context, targetRenderer))
+        return false;
+
+    const auto& svgStyle = style.svgStyle();
+    context.setAlpha(svgStyle.fillOpacity());
+    context.setFillRule(svgStyle.fillRule());
+    context.setFillPattern(m_pattern.copyRef().releaseNonNull());
+    return true;
+}
+
+static AffineTransform transformOnNonScalingStroke(const RenderObject& object, const AffineTransform& resourceTransform)
+{
+    SVGGraphicsElement* element = downcast<SVGGraphicsElement>(object.node());
+    AffineTransform transform = element->getScreenCTM(SVGLocatable::DisallowStyleUpdate);
+    transform *= resourceTransform;
+    return transform;
+}
+
+bool RenderSVGResourcePattern::prepareStrokeOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
+        return false;
+
+    if (!buildPatternIfNeeded(context, targetRenderer))
+        return false;
+
+    const auto& svgStyle = style.svgStyle();
+
+    context.setAlpha(svgStyle.strokeOpacity());
+    SVGRenderSupport::applyStrokeStyleToContext(context, style, targetRenderer);
+    if (svgStyle.vectorEffect() == VectorEffect::NonScalingStroke) {
+        if (auto* shape = dynamicDowncast<RenderSVGShape>(targetRenderer))
+            m_pattern->setPatternSpaceTransform(transformOnNonScalingStroke(targetRenderer, m_pattern->patternSpaceTransform()));
+    }
+    context.setStrokePattern(m_pattern.copyRef().releaseNonNull());
+    return true;
+}
+
+bool RenderSVGResourcePattern::buildTileImageTransform(const RenderElement& renderer, const PatternAttributes& attributes, const SVGPatternElement& patternElement, FloatRect& patternBoundaries, AffineTransform& tileImageTransform) const
+{
+    FloatRect objectBoundingBox = renderer.objectBoundingBox();
```

Comment:
You can fold that in the viewBox case where this variable is needed..

Reply:
But objectBoundingBox is used in the line right after it is defined.

Reply:
Oh missed that. Then only some autos could help here.

Reply:
Done.